### DISCUSSION TITLE
Using Google Fonts with Next.js

### DISCUSSION THREAD

**Discussion by joe-bell** [vercel/next.js#14581] [Show and tell]
Happy Friday Next.js friends 👋 

I've [just published a post](https://joebell.co.uk/fyi/using-google-fonts-with-next-js) detailing my Next.js helper: [**`next-google-fonts`**](https://github.com/joe-bell/next-google-fonts).

It ensures your Google Fonts load **fast** and **asynchronously**.

I've used this approach in a few projects now and it's proved to be pretty successful. If you use Google Fonts, I'd love to hear what you think.

**Comment by TheHolyWaffle** [comment]
What's the advantage of this compared to simply self-hosting google fonts? In my understanding, self-hosting fixes the flash of unstyled text.

Is there less CSS render blocking?

For example, see https://www.npmjs.com/package/typeface-roboto

**Reply by joe-bell** [reply]
Fair question, but I'm not sure I'd agree with "simply self-hosting".

For the absolute-optimum self-hosted fonts you'll need to configure only the things you **need** like character sets and font-weights. Taking a look at that package for example, you'll be bringing **all** the font-weights (which can be configured in Google Fonts), all the characters sets for the font-face, and to top it off you'll still have render-blocking CSS (Webpack isn't treating it as non-critical).

Google's hosted on a CDN and you can tailor font-families and their weights to your needs (plus leverage `font-display: swap;`). I'd argue you get better performance with this than using a self-hosted `typeface-` package, but of course if you have the time to manage character sets and weights manually, you would see a benefit

If that makes sense?

**Reply by TheHolyWaffle** [reply]
It does, there seems to be no reason to not use your approach. Thanks!

**Comment by eric-burel** [comment]
Hi, does your approach avoid Flash of Unstyled Font (FOUT) just after the server-side render? Currently my app displays as Arial before "Source Sans Pro" is loaded from the URL.

**Reply by eric-burel** [reply]
Ok in the Readme you seem to say that we will have a flash of the fallback font so FOUT. Do you think there's a way to get the font in cache? Will it happen on each page reload?

**Reply by eric-burel** [reply]
Eventually I learnt something: you can play around with the `display` option. @joe-bell what is your opinion on a non swap display to avoid any FOUT? Does it really defeat your optimizations? FOUT are really not pleasurable to see but I'd like to keep the website as fast as possible.

**Reply by joe-bell** [reply]
If you choose a non-swap option you will see FOUT

Flash of fallback font is **not** FOUT, that's the fallback font. FOUT is when you don't see any text!

**Reply by eric-burel** [reply]
It is a a Flash of Unstyled Text, so kind of a FOUT if you display the fallback font? Flash of Invisible Text is when you don't see the text. Anyway playing around with the display around it seems that I have to pick one of those 2 scenarios, in my case having a non visible text for a very short amount of time seems more relevant
(edit: all that is my own understanding, I may be completely wrong)
